### PR TITLE
Implement endpoint filter by replication lag.

### PR DESCRIPTION
**Pull request by guoliang100**
@alainjobart @sougou

### CODE REVIEW

File: go/vt/discovery/replicationlag.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	// filter those affecting "mean" lag significantly
+	// calculate mean for all endpoints
+	res := make([]*EndPointStats, 0, len(list))
+	m, err := mean(list, -1)
+	if err != nil {
+		return list
+	}
+	for i, eps := range list {
+		// calculate mean by excluding ith endpoint
+		mi, err := mean(list, i)
```
Comment (alainjobart) [inline go/vt/discovery/replicationlag.go]: list always has 2 elements or more, so mean can never return an error, right? I'd remove the error path for it then.

File: go/vt/discovery/replicationlag.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		}
+		list = append(list, eps)
+	}
+	if len(list) <= 1 {
+		return list
+	}
+	// if all have low replication lag (<=30s), return all endpoints.
+	allLowLag := true
+	for _, eps := range list {
+		if eps.Stats.SecondsBehindMaster > 30 {
```
Comment (alainjobart) [inline go/vt/discovery/replicationlag.go]: That 30 has to be a configurable flag.Duration somewhere.

File: go/vt/discovery/replicationlag_test.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+		Stats:   &qpb.RealtimeStats{SecondsBehindMaster: 120},
+	}
+	got = FilterByReplicationLag([]*EndPointStats{eps1, eps2, eps3, eps4})
+	if len(got) != 3 || !reflect.DeepEqual(got[0], eps1) || !reflect.DeepEqual(got[1], eps2) || !reflect.DeepEqual(got[2], eps3) {
+		t.Errorf("FilterByReplicationLag([5s, 10s, 15s, 120s]) = %+v, want [5s, 10s, 15s]", got)
+	}
+	// lags of (30m, 35m, 40m, 45m)
+	eps1 = &EndPointStats{
+		Serving: true,
+		Stats:   &qpb.RealtimeStats{SecondsBehindMaster: 1800},
```
Comment (alainjobart) [inline go/vt/discovery/replicationlag_test.go]: just to be clearer, I'd replace 1800 with 30*60, and same in all subsequent values.

File: go/vt/discovery/replicationlag_test.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+		Stats:   &qpb.RealtimeStats{},
+	}
+	eps2 := &EndPointStats{
+		Serving: false,
+		Stats:   &qpb.RealtimeStats{},
+	}
+	got = FilterByReplicationLag([]*EndPointStats{eps1, eps2})
+	if len(got) != 1 {
+		t.Errorf("FilterByReplicationLag([EPS{Serving: true}, EPS{Serving: false}]) = %+v, want one", got)
+	}
```
Comment (alainjobart) [inline go/vt/discovery/replicationlag_test.go]: Might want to set a unique UID in eps, and make sure you get the right result here.

File: go/vt/discovery/replicationlag_test.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		Stats:   &qpb.RealtimeStats{SecondsBehindMaster: 1},
+	}
+	eps4 := &EndPointStats{
+		Serving: true,
+		Stats:   &qpb.RealtimeStats{SecondsBehindMaster: 30},
+	}
+	got = FilterByReplicationLag([]*EndPointStats{eps1, eps2, eps3, eps4})
+	if len(got) != 4 || !reflect.DeepEqual(got[0], eps1) || !reflect.DeepEqual(got[1], eps2) || !reflect.DeepEqual(got[2], eps3) || !reflect.DeepEqual(got[3], eps4) {
+		t.Errorf("FilterByReplicationLag([1s, 1s, 1s, 30s]) = %+v, want all", got)
+	}
```
Comment (alainjobart) [inline go/vt/discovery/replicationlag_test.go]: again, set a UID in eps here. To make sure you don't get 3x eps1, that happens to be the same as eps2 and eps3.

**Comment by alainjobart** [comment]
LGTM after comments are addressed.

File: go/vt/discovery/replicationlag.go (hunk 1 above)
Comment (guoliang100) [inline go/vt/discovery/replicationlag.go]: Done

File: go/vt/discovery/replicationlag.go (hunk 2 above)
Comment (guoliang100) [inline go/vt/discovery/replicationlag.go]: Done. SecondsBehindMaster is uint32 (not Duration), so I created a flag as Uint.

File: go/vt/discovery/replicationlag_test.go (hunk 3 above)
Comment (guoliang100) [inline go/vt/discovery/replicationlag_test.go]: Done

File: go/vt/discovery/replicationlag_test.go (hunk 4 above)
Comment (guoliang100) [inline go/vt/discovery/replicationlag_test.go]: Done

File: go/vt/discovery/replicationlag_test.go (hunk 5 above)
Comment (guoliang100) [inline go/vt/discovery/replicationlag_test.go]: Done

File: go/vt/discovery/replicationlag.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
+package discovery
+
+import (
+	"flag"
+	"fmt"
+)
+
+var replicationLowLag = flag.Uint("discovery_replication_low_lag", 30, "the replication lag in seconds that is considered low")
```
Comment (alainjobart) [inline go/vt/discovery/replicationlag.go]: Please make this a flag.Duration, and when using it, just use .Seconds. That way the command line parsing will work for 20s, 1m or any other shortcut, and will be cleaner.

(the fact that we round it to the second is an implementation detail, and we may get more granular at some point).

Comment (guoliang100) [inline go/vt/discovery/replicationlag.go]: Done

**Comment by alainjobart** [comment]
LGTM

**Pull request merged**